File: content/tracks/dsa-python/concepts/16-backtracking/backtracking_test.py
import sys
from stub import subsets, permutations, combination_sum

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


def normalize(lol: list[list[int]]) -> list[list[int]]:
    """Sort each inner list, then sort the outer list for order-independent comparison."""
    return sorted(sorted(inner) for inner in lol)


# subsets tests
result = normalize(subsets([1, 2, 3]))
expected = normalize([[], [1], [2], [3], [1, 2], [1, 3], [2, 3], [1, 2, 3]])
test("subsets: [1,2,3]", result == expected, f"got {result}")

result = normalize(subsets([]))
test("subsets: empty input", result == [[]], f"got {result}")

result = normalize(subsets([1]))
test("subsets: single element", result == [[], [1]], f"got {result}")

result = normalize(subsets([0]))
test("subsets: single zero", result == [[], [0]], f"got {result}")

# Check count only for larger input
result = subsets([1, 2, 3, 4])
test("subsets: [1,2,3,4] count", len(result) == 16, f"got {len(result)} subsets, want 16")

# permutations tests
result = normalize(permutations([1, 2, 3]))
expected = normalize([[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]])
test("perms: [1,2,3]", result == expected, f"got {result}")

result = normalize(permutations([1]))
test("perms: single element", result == [[1]], f"got {result}")

result = normalize(permutations([]))
test("perms: empty", result == [[]], f"got {result}")

result = permutations([1, 2, 3, 4])
test("perms: [1,2,3,4] count", len(result) == 24, f"got {len(result)}, want 24")

# combination_sum tests
result = normalize(combination_sum([2, 3, 6, 7], 7))
expected = normalize([[2, 2, 3], [7]])
test("combo_sum: [2,3,6,7] target=7", result == expected, f"got {result}")

result = normalize(combination_sum([2, 3, 5], 8))
expected = normalize([[2, 2, 2, 2], [2, 3, 3], [3, 5]])
test("combo_sum: [2,3,5] target=8", result == expected, f"got {result}")

result = combination_sum([2], 1)
test("combo_sum: impossible", result == [], f"got {result}")

result = normalize(combination_sum([1], 3))
test("combo_sum: single coin", result == [[1, 1, 1]], f"got {result}")

result = normalize(combination_sum([2, 4], 6))
expected = normalize([[2, 2, 2], [2, 4]])
test("combo_sum: [2,4] target=6", result == expected, f"got {result}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)

File: content/tracks/dsa-typescript/concepts/16-backtracking/backtracking_test.ts
import { subsets, permutations, combinationSum } from "./stub";

let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    console.log(`  FAIL: ${name} — ${msg}`);
    failed++;
  }
}

function assertEqual<T>(actual: T, expected: T): void {
  if (actual !== expected) {
    throw new Error(`expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`);
  }
}

function sortedSubsets(arr: number[][]): string {
  return JSON.stringify(
    arr.map((s) => [...s].sort((a, b) => a - b)).sort((a, b) => {
      if (a.length !== b.length) return a.length - b.length;
      return JSON.stringify(a) < JSON.stringify(b) ? -1 : 1;
    })
  );
}

// subsets tests
test("subsets: [1,2,3] has 8 subsets", () => {
  assertEqual(subsets([1, 2, 3]).length, 8);
});

test("subsets: includes empty set", () => {
  const result = subsets([1, 2, 3]);
  const hasEmpty = result.some((s) => s.length === 0);
  if (!hasEmpty) throw new Error("missing empty subset");
});

test("subsets: includes full set", () => {
  const result = subsets([1, 2, 3]);
  const hasFull = result.some((s) => s.length === 3);
  if (!hasFull) throw new Error("missing full subset");
});

test("subsets: no duplicates", () => {
  const result = subsets([1, 2, 3]);
  const strs = result.map((s) => JSON.stringify([...s].sort()));
  const unique = new Set(strs);
  assertEqual(unique.size, 8);
});

test("subsets: empty input", () => {
  assertEqual(subsets([]).length, 1); // just the empty set
});

test("subsets: single element", () => {
  assertEqual(subsets([5]).length, 2);
});

// permutations tests
test("permutations: [1,2,3] has 6 permutations", () => {
  assertEqual(permutations([1, 2, 3]).length, 6);
});

test("permutations: all have correct length", () => {
  const result = permutations([1, 2, 3]);
  const allCorrectLength = result.every((p) => p.length === 3);
  if (!allCorrectLength) throw new Error("some permutation has wrong length");
});

test("permutations: no duplicates", () => {
  const result = permutations([1, 2, 3]);
  const strs = result.map((p) => JSON.stringify(p));
  const unique = new Set(strs);
  assertEqual(unique.size, 6);
});

test("permutations: single element", () => {
  assertEqual(permutations([1]).length, 1);
});

// combinationSum tests
test("combinationSum: [2,3,6,7] target=7", () => {
  const result = combinationSum([2, 3, 6, 7], 7);
  assertEqual(sortedSubsets(result), sortedSubsets([[2, 2, 3], [7]]));
});

test("combinationSum: [2,3] target=6", () => {
  const result = combinationSum([2, 3], 6);
  assertEqual(sortedSubsets(result), sortedSubsets([[2, 2, 2], [3, 3]]));
});

test("combinationSum: no solution", () => {
  assertEqual(combinationSum([3], 5).length, 0);
});

test("combinationSum: target=0", () => {
  // No candidates needed; only the empty combination
  assertEqual(combinationSum([1, 2], 0).length, 1);
});

test("combinationSum: single candidate", () => {
  const result = combinationSum([2], 4);
  assertEqual(sortedSubsets(result), sortedSubsets([[2, 2]]));
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);
